architecture/crds.md

```diff
@@ -0,0 +1,324 @@
+
+As we progress through potential solutions, keep in mind the roles and caveats elaborated before:
+
+- Chart authors create charts
+- Helm users install and manage instances of those charts
+- We cannot assume that those roles are filled by the same person
+- We cannot assume, specifically, that a Helm user understands the inner workings of a chart or even of Kubernetes
+- We cannot assume that a Helm user looks at the contents of a chart
+- We cannot assume that a chart author knows the target cluster environment
+- We cannot assume that a chart author knows what other things will be running concurrently in the same cluster as the chart
+- We NEVER want to assert that it's "the users problem" to figure out when a chart operation (install, upgrade, delete, rollback) could cause a catastrophic cluster event reaching beyond the specific chart installation.
+
+The following kinds of solutions have been suggested:
+
+1. Do not manage CRDs any differently than other resources
+2. Helm should let chart authors control how CRDs are managed
+3. Helm should let Helm users control how CRDs are managed
+4. Helm should have deep special-case logic for handling CRDs
+
+Each of these is discussed below.
+
+### No Special CRD Management
+
+In this solution, a CRD is treated no differently than any other Kubernetes kind.
+
+_Prima facie_, this solution sounds like a no-nonsense go-to. Just drop CRDs in like any other resource and let the world figure out how to do this.
+
+This was our initial approach, and it worked fine in cases where all a chart did was install the CRD and possibly the controller that consumed that CRD. In other words, it worked _only for_ the original CRD case for a global CRD and a single global controller for that CRD.
+
+But this strategy did not scale as CRD usage followed new patterns. Earlier in this document, we enumerated different CRD patterns that have emerged. Helm's original design does not work as soon as CRDs break out of the simple case. It failed for the following reasons:
+
+1. CRDs alter the behavior of the API server. Specifically, discovery and validation change when a CRD is added or modified.
+2. It is impossible to pre-verify a chart that has both a CRD and CRs for that CRD
+3. `helm delete` becomes a very, very dangerous command, as it can destabilize a cluster by deleting a CRD
+4. `helm upgrade` and `helm rollback` can have surprising consequences that are difficult to find and debug
+
+We still haven't solved #2, but by adding some special treatment of CRDs in Helm 3, we eliminated #3-4 and fixed #1.
+
+> **IMPORTANT**: It is still possible to use Helm this way. For CRD/Operator-only charts, this is a perfectly legitimate way of packaging.
```

Comment:
Is this true of charts that are only intended to be installed once-per-cluster, e.g., [cert-manager](https://github.com/helm/helm/issues/7735) or Agones? Otherwise, you're still sharing CRDs between multiple Helm releases, e.g. [Jaeger Operator](https://github.com/jaegertracing/jaeger-operator), and all the same issues appear.

If so, perhaps refer to "CRD and Singleton Controller" pattern mentioned in the earlier survey, as "CRD/Operator-only" is not clearly defined here.

Reply:
> Is this true of charts that are only intended to be installed once-per-cluster

This is where things get difficult. There is the intent of the Chart creator and then there are situations where it doesn't work out for a variety of reasons. The chart creators does not control the chart installer. I've seen cases such as...

- Someone is doing dev and staging in the same cluster... dealing with two versions... even accidentally screwing it up
- Two different tenants trying to install something and those tenants don't know about each other. To assume it's only installed once it to assume that organization has certain controls in place that may or may not be there in reality.

Problems arise when we break from the happy path. That happens in reality fairly often and with people who may not be experts and can't figure out what broke easily.

Reply:
I think maybe I wasn't clear in my point. I was trying to say that using Helm "this way" (i.e. packaging the CRDs as regular resources managed by the Helm release) is only really legitimate for once-per-cluster charts, the "CRD and Singleton Controller" pattern mentioned earlier. Perhaps that's what "CRD/Operator-only" charts meant, but it wasn't clear to me.

If a multi-install chart treats CRDs as regular resources, the first tenant to install the chart now owns the CRs of every other tenant to install that chart (who have to install with CRDs disabled, assuming there's a template value to do that), and if the first tenant uninstalls first, everyone else loses their data.

Whether "once-per-cluster" charts are a good idea is a separate issue, as you've outlined in your comments. I happen to use two of them in my work, but I also have taken to managing their CRDs using `kubectl apply` _because_ they handle CRDs are regular objects, and I've definitely corrupted Agones installations a few times during upgrades because of it.